llvm/lib/SYCLLowerIR/SYCLJointMatrixTransform.cpp

```diff
@@ -22,6 +22,74 @@ namespace {
 static constexpr char ACCESS_CHAIN[] = "_Z19__spirv_AccessChain";
 static constexpr char MATRIX_TYPE[] = "spirv.CooperativeMatrixKHR";
 
+Type *getInnermostType(Type *Ty) {
+  while (auto *ArrayTy = dyn_cast<ArrayType>(Ty))
+    Ty = ArrayTy->getElementType();
+  return Ty;
+}
+
+Type *replaceInnermostType(Type *Ty, Type *NewInnermostTy) {
+  if (auto *ArrayTy = dyn_cast<ArrayType>(Ty))
```

Comment:
Ideally we should also handle `struct Example { sycl::joint_matrix(...) };` but it goes outside of the scope of this PR.
Also some weird things can appear in a case of the unions, while it strange to use them for joint matrices as well as for any opaque types, I don't remember we disallow such case.

Reply:
Makes sense. Goal of this PR was to address crash in perf test. Feel free to extend the code to cover more cases.